### PR TITLE
avoid race condition in bootstrap certs for parallel runs

### PR DESCRIPTION
**Pull request by azhao155**
This pr is create per cluster certs for docker-env. So in parallel testing, would fix race condition issue of getting certs.

### CODE REVIEW

**Comment by azhao155** [comment]
send out pr to verify if it can fix the flakiness.

File: pkg/provision/provision.go [hunk 1]
```diff
@@ -102,7 +103,11 @@ func configureAuth(p miniProvisioner) error {
 		return errors.Wrap(err, "error getting ssh hostname during provisioning")
 	}
 
-	if err := copyHostCerts(authOptions); err != nil {
+	//if err := copyHostCerts(authOptions); err != nil {
```
Comment (medyagh) [inline pkg/provision/provision.go:106]: any reason we have this commented code?

**Review by medyagh — CHANGES_REQUESTED** [review]

**Comment by medyagh** [comment]
On first test seems to be fixing the docker-env flake

File: pkg/minikube/machine/client.go [hunk 2]
```diff
@@ -71,6 +72,7 @@ func NewAPIClient(miniHome ...string) (libmachine.API, error) {
 		storePath:    storePath,
 		Filestore:    persist.NewFilestore(storePath, certsDir, certsDir),
 		legacyClient: NewRPCClient(storePath, certsDir),
+		flock:        fslock.New(localpath.MakeMiniPath("fileLock.txt")),
```
Comment (medyagh) [inline pkg/minikube/machine/client.go:75]: let give a better name to the lock,how about "machine_client.lock"  ?

File: pkg/minikube/machine/client.go [hunk 3]
```diff
@@ -183,7 +186,21 @@ func (api *LocalClient) Create(h *host.Host) error {
 	}{
 		{
 			"bootstrapping certificates",
-			func() error { return cert.BootstrapCertificates(h.AuthOptions()) },
+			func() error {
+				// CA cert and client cert should be generated atomically, otherwise might cause bad certificate error
+				lockErr := api.flock.LockWithTimeout(time.Second * 5)
+				if lockErr != nil {
+					return fmt.Errorf("falied to acquire lock > " + lockErr.Error())
```
Comment (medyagh) [inline pkg/minikube/machine/client.go:193]: use %v formatting for the error and maybe more descriptive
```
fmt.Errorf("failed to acquire bootstrap client lock: %v " + lockErr)
```

File: pkg/minikube/machine/client.go [hunk 4]
```diff
@@ -183,7 +186,21 @@ func (api *LocalClient) Create(h *host.Host) error {
...
+				}
+				defer func() {
+					lockErr = api.flock.Unlock()
+					if lockErr != nil {
+						klog.Errorf("falied to release lock > %s", lockErr.Error())
```
Comment (medyagh) [inline pkg/minikube/machine/client.go:198]: same here, use %v
```
klog.Errorf("falied to release bootstrap cert client lock  %v", lockErr)

```

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/minikube/machine/client.go [hunk 5]
```diff
@@ -183,7 +186,21 @@ func (api *LocalClient) Create(h *host.Host) error {
...
+			func() error {
+				// CA cert and client cert should be generated atomically, otherwise might cause bad certificate error
+				lockErr := api.flock.LockWithTimeout(time.Second * 5)
+				if lockErr != nil {
+					return fmt.Errorf("failed to acquire bootstrap client lock: %v " + lockErr.Error())
+				}
+				defer func() {
+					lockErr = api.flock.Unlock()
+					if lockErr != nil {
+						klog.Errorf("falied to release bootstrap cert client lock  %v", lockErr.Error())
```
Comment (medyagh) [inline pkg/minikube/machine/client.go:198]: colon here too

klog.Errorf("falied to release bootstrap cert client lock:  %v")

File: pkg/minikube/machine/client.go [hunk 6]
```diff
@@ -183,7 +186,21 @@ func (api *LocalClient) Create(h *host.Host) error {
 	}{
 		{
 			"bootstrapping certificates",
-			func() error { return cert.BootstrapCertificates(h.AuthOptions()) },
+			func() error {
+				// CA cert and client cert should be generated atomically, otherwise might cause bad certificate error
+				lockErr := api.flock.LockWithTimeout(time.Second * 5)
```
Comment (medyagh) [inline pkg/minikube/machine/client.go:191]: add issue link here why we do this...say this causes docker-env in parallel work

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/minikube/machine/client.go [hunk 7]
```diff
@@ -183,7 +186,23 @@ func (api *LocalClient) Create(h *host.Host) error {
 	}{
 		{
 			"bootstrapping certificates",
-			func() error { return cert.BootstrapCertificates(h.AuthOptions()) },
+			func() error {
+				// Due to https://github.com/docker/machine/issues/3634, we need to have lock for certs generation, otherwise
+				// it will break Docker-Env test in parallel tesing.
```
Comment (medyagh) [inline pkg/minikube/machine/client.go:191]: add the issue number on github 
and change wording to
"this lock is needed to run docker-env tests in parallel issuue #... "

**Review by medyagh — CHANGES_REQUESTED** [review]

**Review by medyagh — APPROVED** [review]

**Pull request merged**